modules/geo-layers/src/tile-3d-layer/get-frame-state.js

```diff
@@ -0,0 +1,96 @@
+const scratchPosition = new Vector3();
+const cullingVolume = new CullingVolume([
+  new Plane(),
+  new Plane(),
+  new Plane(),
+  new Plane(),
+  new Plane(),
+  new Plane()
+]);
+
+// Extracts a frame state appropriate for tile culling from a deck.gl viewport
+// TODO - this could likely be generalized and merged back into deck.gl for other culling scenarios
+export function getFrameState(viewport, animationProps) {
+  // Traverse and and request. Update _selectedTiles so that we know what to render.
+  const {height, tick} = animationProps;
+  const {cameraDirection, cameraUp} = viewport;
+  const {metersPerPixel} = viewport.distanceScales;
+
+  const viewportCenterCartographic = [viewport.longitude, viewport.latitude, 0];
+  // TODO - Ellipsoid.eastNorthUpToFixedFrame() breaks on raw array, create a Vector.
+  // TODO - Ellipsoid.eastNorthUpToFixedFrame() takes a cartesian, is that intuitive?
+  const viewportCenterCartesian = Ellipsoid.WGS84.cartographicToCartesian(
+    viewportCenterCartographic,
+    new Vector3()
+  );
+  const enuToFixedTransform = Ellipsoid.WGS84.eastNorthUpToFixedFrame(viewportCenterCartesian);
+
+  const cameraPositionCartographic = viewport.unprojectPosition(viewport.cameraPosition);
+  const cameraPositionCartesian = Ellipsoid.WGS84.cartographicToCartesian(
+    cameraPositionCartographic,
+    new Vector3()
+  );
+
+  // These should still be normalized as the transform has scale 1 (goes from meters to meters)
+  const cameraDirectionCartesian = new Vector3(
+    enuToFixedTransform.transformAsVector(new Vector3(cameraDirection).scale(metersPerPixel))
+  ).normalize();
+  const cameraUpCartesian = new Vector3(
+    enuToFixedTransform.transformAsVector(new Vector3(cameraUp).scale(metersPerPixel))
+  ).normalize();
+
+  commonSpacePlanesToWGS84(viewport);
+
+  // TODO: make a file/class for frameState and document what needs to be attached to this so that traversal can function
+  return {
+    camera: {
+      position: cameraPositionCartesian,
+      direction: cameraDirectionCartesian,
+      up: cameraUpCartesian
+    },
+    height,
+    cullingVolume,
+    frameNumber: tick, // TODO: This can be the same between updates, what number is unique for between updates?
+    sseDenominator: 1.15 // Assumes fovy = 60 degrees
+  };
+}
+
+function commonSpacePlanesToWGS84(viewport) {
+  // Extract frustum planes based on current view.
+  const viewportCenterCartographic = [viewport.longitude, viewport.latitude, 0];
+  const viewportCenterCartesian = Ellipsoid.WGS84.cartographicToCartesian(
+    viewportCenterCartographic,
+    new Vector3()
+  );
+
+  const frustumPlanes = viewport.getFrustumPlanes();
+  let i = 0;
+  for (const dir in frustumPlanes) {
+    const plane = frustumPlanes[dir];
+    const distanceToCenter = plane.normal.dot(viewport.center);
+    const nLen = plane.normal.len();
+    scratchPosition
+      .copy(plane.normal)
+      .scale((plane.distance - distanceToCenter) / (nLen * nLen))
```

Comment:
I think `plane.normal` has been fixed (normalized) so `nLen * nLen` can be removed.